lib/bootsnap/cli/worker_pool.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+module Bootsnap
+  class CLI
+    class WorkerPool
+      class << self
+        def create(size:, jobs:)
+          if size > 0 && Process.respond_to?(:fork)
+            new(size: size, jobs: jobs)
+          else
+            Inline.new(jobs: jobs)
+          end
+        end
+      end
+
+      class Inline
+        def initialize(jobs: {})
+          @jobs = jobs
+        end
+
+        def push(job, *args)
+          @jobs.fetch(job).call(*args)
+          nil
+        end
+
+        def spawn
+          # noop
+        end
+
+        def shutdown
+          # noop
+        end
+      end
+
+      class Worker
+        attr_reader :pipe, :pid
+
+        def initialize(jobs)
+          @jobs = jobs
+          @pipe_out, @pipe = IO.pipe
+          @pid = nil
+        end
+
+        def work_loop
+          loop do
+            job, *args = Marshal.load(@pipe_out)
+            return if job == :exit
+            @jobs.fetch(job).call(*args)
+          end
+        rescue IOError
+          nil
+        end
+
+        def spawn
+          @pid = Process.fork do
+            @pipe.close
+            work_loop
+            exit!(0)
+          end
+          @pipe_out.close
+          true
+        end
+      end
+
+      def initialize(size:, jobs: {})
+        @size = size
+        @jobs = jobs
+        @queue = Queue.new
+        @pids = []
+      end
+
+      def spawn
+        @workers = @size.times.map { Worker.new(@jobs) }
+        @workers.each(&:spawn)
+        @dispatcher_thread = Thread.new { dispatch_loop }
+        @dispatcher_thread.abort_on_exception = true
+        true
+      end
+
+      def dispatch_loop
+        loop do
+          case job = @queue.pop
+          when nil
+            @workers.each do |worker|
+              worker.pipe.write(Marshal.dump([:exit]))
+              worker.pipe.close
+            end
+            return true
+          else
+            free_pipe.write(Marshal.dump(job))
+          end
+        end
+      end
+
+      def free_pipe
+        IO.select(nil, @workers.map(&:pipe).shuffle)[1].first
```

Comment:
> We could implement `Worker#to_io `

TIL, that's an awesome API! Thanks!

> I'm curious about the shuffle, is that required to avoid getting the same worker each time?

Yeah, `IO.select` returns the IOs in the same order, so without some randomness the first workers would get all the messages.

> we could maybe use `sample`

Good point.